### PR TITLE
feat(account-settings): Add ConfigureTOTP component

### PR DESCRIPTION
**Pull request by wlee221**
#### Description of changes

This adds `ConfigureTOTP` component.
- functionality for configuring totp when mfa is optional
- prop support for `error/success` handlers, plus custom `totpUsername` and `totpIssuer`
- Default components for each part of the UI

Component override is **not** implemented yet. This will happen after functionalities are done for account settings.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are updated
- [ ] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: examples/next/pages/ui/components/account-settings/configure-totp/index.page.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    <Authenticator>
+      {({ signOut }) => (
+        <Card width="800px">
+          <Flex direction="column">
+            <Card variation="outlined">
+              <Flex direction="column">
+                <Heading>Setup MFA:</Heading>
+                <ConfigureTOTP
+                  onSuccess={() => {
+                    console.log('+++');
```
Comment (wlee221) [inline examples/next/pages/ui/components/account-settings/configure-totp/index.page.tsx:32]:
```suggestion
```

File: examples/next/pages/ui/components/account-settings/configure-totp/index.page.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import '@aws-amplify/ui-react/styles.css';
+
+import { Alert, ConfigureTOTP } from '@aws-amplify/ui-react';
+
+import awsExports from './aws-exports';
+Amplify.configure(awsExports);
+
+export default function App() {
+  const [isSuccessful, setIsSuccessful] = React.useState(false);
+  console.log(isSuccessful);
```
Comment (wlee221) [inline examples/next/pages/ui/components/account-settings/configure-totp/index.page.tsx:21]:
```suggestion
```

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+      try {
+        const newSecretKey = await setupTOTP(user);
+        const username = totpUsername || user?.username;
+        const totpCode = getTotpCodeURL(totpIssuer, username, newSecretKey);
+        const qrCodeImageSource = await QRCode.toDataURL(totpCode);
+
+        if (!hasInit.current) {
+          setSecretKey(newSecretKey);
+          setQrCode(qrCodeImageSource);
+          hasInit.current = false;
```
Comment (calebpollman) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:53]: Should this be setting `hasInit.current` to `true`?

File: examples/next/pages/ui/components/account-settings/configure-totp/index.page.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+              <Flex direction="column">
+                <Heading>Setup MFA:</Heading>
+                <ConfigureTOTP
+                  onSuccess={() => {
+                    setIsSuccessful(true);
+                  }}
+                />
+                {isSuccessful ? (
+                  <Alert variation="success" isDismissible>
+                    TOTP has been set up successfully
```
Comment (joebuono) [inline examples/next/pages/ui/components/account-settings/configure-totp/index.page.tsx:36]: Do we need translations for these strings?

File: examples/next/pages/ui/components/account-settings/configure-totp/index.page.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+export default function App() {
+  const [isSuccessful, setIsSuccessful] = React.useState(false);
+  return (
+    <Authenticator>
+      {({ signOut }) => (
+        <Card width="800px">
+          <Flex direction="column">
+            <Card variation="outlined">
+              <Flex direction="column">
+                <Heading>Setup MFA:</Heading>
```
Comment (joebuono) [inline examples/next/pages/ui/components/account-settings/configure-totp/index.page.tsx:28]: Do we need translations for these strings?

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx [hunk 6]
```diff
@@ -0,0 +1,167 @@
+import React from 'react';
+import QRCode from 'qrcode';
+
+import {
+  setupTOTP,
```
Comment (joebuono) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:5]: nit: alphabetize

Comment (wlee221) [reply packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx]: Done in [6fee5d1](https://github.com/aws-amplify/amplify-ui/pull/2969/commits/6fee5d15f7c6f6bdbdf9e1d572cb4c91bd6f4700)!

File: packages/react/src/components/AccountSettings/ConfigureTOTP/types.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+export interface ConfigureTOTPProps {
+  /** TOTP issuer. Default to "AWSCognito" */
+  totpIssuer?: string;
+  /** TOTP username. Default to  */
```
Comment (joebuono) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/types.ts:4]: What should it default to?

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        }
+      } catch (e) {
+        logger.error(e);
+      }
+    },
+    [totpIssuer, totpUsername]
+  );
+
+  React.useEffect(() => {
+    if (user && !hasInit.current) {
```
Comment (calebpollman) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:63]: Is there a scenario where `user` is falsy?

File: packages/react/src/components/AccountSettings/ConfigureTOTP/defaults.tsx [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import {
+  AccountSettingsError,
+  AccountSettingsImage,
+  AccountSettingsSubmitButton,
+  AccountSettingsTextField,
+} from '../types';
+
+export const ConfirmationCode: AccountSettingsTextField = TextField;
+
+export const SecretKeyQRCode: AccountSettingsImage = Image;
```
Comment (joebuono) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/defaults.tsx:13]: I like how this makes it easier to understand when just looking at the `ConfigureTOTP` component 👍

**Review by joebuono — APPROVED** [review]

File: packages/react/src/components/AccountSettings/ConfigureTOTP/types.ts (hunk 7 above)
Comment (wlee221) [reply packages/react/src/components/AccountSettings/ConfigureTOTP/types.ts:4]: Oh man, I screwed this up didn't I 😓 

it defaults to `user.username`, which is taken from the `user` object.
```suggestion
  /** TOTP username. Default to `user.username`  */
```

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+  return (
+    <View as="form" onSubmit={handleSubmit}>
+      <Flex direction="column" alignItems="center">
+        {qrCode ? (
+          <SecretKeyQRCode
+            data-amplify-qrcode
+            src={qrCode}
+            alt="qr code"
+            width="228px"
+            height="228px"
```
Comment (calebpollman) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:136]: We should put these in a constant maybe? For example:
```
const QR_CODE_SIZE = "228px";
const QR_CODE_DIMENSIONS = {
  height: QR_CODE_SIZE,
  width: QR_CODE_SIZE,
}
```

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+  if (isLoading) {
+    return null;
+  }
+
+  return (
+    <View as="form" onSubmit={handleSubmit}>
+      <Flex direction="column" alignItems="center">
+        {qrCode ? (
+          <SecretKeyQRCode
+            data-amplify-qrcode
```
Comment (calebpollman) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:132]: Unless we really need to, let's avoid using `data-` attributes

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+        <ConfirmationCode
+          alignSelf="stretch"
+          isRequired
+          label="Confirmation Code"
+          name="code"
+          onChange={handleChange}
+          placeholder="Code"
+          value={formValues.code}
+          width="100%"
+        ></ConfirmationCode>
```
Comment (calebpollman) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:151]:
```suggestion
        />
```

File: packages/react/src/hooks/useAuth.ts [hunk 13]
```diff
@@ -28,7 +28,7 @@ export const useAuth = (): UseAuthResult => {
    * This util will be used to get current user after those events.
    */
   const fetchCurrentUser = React.useCallback(async () => {
-    setResult({ user: undefined, isLoading: true, error: undefined });
+    setResult((result) => ({ ...result, isLoading: true }));
```
Comment (calebpollman) [inline packages/react/src/hooks/useAuth.ts:31]: To prevent confusion on the future with the `result` declared on [line 20](https://github.com/aws-amplify/amplify-ui/pull/2969/files#diff-20eb679d6cd90516423c4c8bec245d4b9d21666124c29bd07a2ecbcd7c1feb93R20), and to make it clear that we are referencing the previous value of `result`, suggest `prevResult`

File: packages/react/src/components/AccountSettings/ConfigureTOTP/types.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+  totpUsername?: string;
+  // callback once totp is successfully set up
```
Comment (calebpollman) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/types.ts:6]: Sorry to keep on abt comment syntaxes, but can we align on either inline or JSDoc for this interface?

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+        </CopySecretKey>
+        <ConfirmationCode
+          alignSelf="stretch"
+          isRequired
+          label="Confirmation Code"
+          name="code"
+          onChange={handleChange}
+          placeholder="Code"
+          value={formValues.code}
+          width="100%"
```
Comment (calebpollman) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:150]: Just curious, do we need both `alignSelf="stretch"` and `width="100%"`?

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx (hunk 8 above)
Comment (wlee221) [reply packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:63]: No, it is an assumption that `user` is already authenticated by the time customer runs this component. 

If not, `useAuth` and early return by logging error: https://github.com/aws-amplify/amplify-ui/blob/a62a1f08d7eaad99ce6909dc0ae72eb0939ffdb1/packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx#L117-L120

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx (hunk 11 above)
Comment (wlee221) [reply packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:132]: Agreed, it's not like it has its own style anyway

Comment (wlee221) [reply packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:132]:
```suggestion
```

File: packages/react/src/hooks/useAuth.ts (hunk 13 above)
Comment (wlee221) [reply packages/react/src/hooks/useAuth.ts:31]: Makes sense,

```suggestion
    setResult((prevResult) => ({ ...prevResult, isLoading: true }));
```

File: packages/react/src/components/AccountSettings/ConfigureTOTP/types.ts (hunk 14 above)
Comment (wlee221) [reply packages/react/src/components/AccountSettings/ConfigureTOTP/types.ts:6]: No, this is my bad on sloppiness here xD

```suggestion
  /** callback once totp is successfully set up */
```

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx (hunk 15 above)
Comment (wlee221) [reply packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:150]: We do not, good catch.
```suggestion
```

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx (hunk 3 above)
Comment (wlee221) [reply packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx]: 100%, this is a typo from my end. [b48ab69](https://github.com/aws-amplify/amplify-ui/pull/2969/commits/b48ab6988dbae259e3ed25e3a0dc3bc4d04bbec9)

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx (hunk 10 above)
Comment (wlee221) [reply packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx]: Agreed, done in [bccefa0](https://github.com/aws-amplify/amplify-ui/pull/2969/commits/bccefa086bc54d32bb6abd2c4e879d651f5e2970).

File: packages/react/src/components/AccountSettings/types.ts [hunk 16]
```diff
@@ -12,17 +14,27 @@ import {
...
  */
 export type AccountSettingsPasswordField<Props = {}> = React.ComponentType<
   // `Props` generic allows additional props passed on override components
   Props & CommonPasswordFieldProps & { validationErrors?: string[] }
 >;
 
+export type AccountSettingsTextField<Props = {}> = React.ComponentType<
+  Props & CommonTextFieldProps
+>;
+
```
Comment (ErikCH) [inline packages/react/src/components/AccountSettings/types.ts:33]: I like this pattern, I've been using it with File Uploader

File: packages/react/src/hooks/useAuth.ts [hunk 17]
```diff
@@ -28,7 +28,7 @@ export const useAuth = (): UseAuthResult => {
    * This util will be used to get current user after those events.
    */
   const fetchCurrentUser = React.useCallback(async () => {
-    setResult({ user: undefined, isLoading: true, error: undefined });
+    setResult((prevResult) => ({ ...prevResult, isLoading: true }));
```
Comment (ErikCH) [inline packages/react/src/hooks/useAuth.ts:31]: Why this change?

**Review by ErikCH — APPROVED** [review]
Looks good, just one question

File: packages/react/src/hooks/useAuth.ts (hunk 17 above)
Comment (wlee221) [reply packages/react/src/hooks/useAuth.ts:31]: Oh yeah, that needed some comment!

So this prevents `useAuth` from losing `user` value between `fetchCurrentUser` calls, by always spreading the previous value onto the new result.

File: packages/react/src/components/AccountSettings/types.ts (hunk 16 above)
Comment (wlee221) [reply packages/react/src/components/AccountSettings/types.ts:33]: It's pretty clever!

File: packages/react/src/components/AccountSettings/ConfigureTOTP/types.ts [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+export interface ConfigureTOTPProps {
+  /** TOTP issuer. Default to "AWSCognito" */
+  totpIssuer?: string;
+  /** TOTP username. Default to `user.username`  */
+  totpUsername?: string;
+  /** callback once totp is successfully set up */
+  onSuccess?: () => void;
+  // callback when there's an error
```
Comment (calebpollman) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/types.ts:18]:
```suggestion
  /** callback when there's an error */
```

File: packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+  // translations
+  const confirmText = translate('Confirm');
+  const copyCodeText = translate('Copy Secret Code');
+
+  // event handlers
+  const handleChange = React.useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      event.preventDefault();
+      const { name, value } = event.target;
+      setFormValues((formValues) => ({ ...formValues, [name]: value }));
```
Comment (calebpollman) [inline packages/react/src/components/AccountSettings/ConfigureTOTP/ConfigureTOTP.tsx:75]:
```suggestion
      setFormValues((prevFormValues) => ({ ...prevFormValues, [name]: value }));
```
